flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/compiler/NFACompiler.java

```diff
@@ -657,25 +663,34 @@ private boolean isPatternOptional(Pattern<T, ?> pattern) {
 				true);
 
 			IterativeCondition<T> proceedCondition = getTrueFunction();
-			if (currentPattern.getQuantifier().isGreedy()) {
-				proceedCondition = getGreedyCondition(proceedCondition, Lists.newArrayList(takeCondition));
+			if (currentPattern.getQuantifier().hasProperty(Quantifier.QuantifierProperty.GREEDY)) {
+				proceedCondition = getGreedyCondition(
+					proceedCondition,
+					takeCondition,
+					ignoreCondition,
+					followingTakeCondition);;
 			}
 			final State<T> loopingState = createState(currentPattern.getName(), State.StateType.Normal,
-				currentPattern.getQuantifier().isGreedy());
+				currentPattern.getQuantifier().hasProperty(Quantifier.QuantifierProperty.GREEDY));
 			loopingState.addProceed(sinkState, proceedCondition);
 			loopingState.addTake(takeCondition);
 
 			addStopStateToLooping(loopingState);
 
 			if (ignoreCondition != null) {
 				final State<T> ignoreState = createState(currentPattern.getName(), State.StateType.Normal,
-					currentPattern.getQuantifier().isGreedy());
+					currentPattern.getQuantifier().hasProperty(Quantifier.QuantifierProperty.GREEDY));
 				ignoreState.addTake(loopingState, takeCondition);
 				ignoreState.addIgnore(ignoreCondition);
 				loopingState.addIgnore(ignoreState, ignoreCondition);
+				if (currentPattern.getQuantifier().hasProperty(Quantifier.QuantifierProperty.GREEDY)) {
```

Comment:
I think the problem you mentioned with `GreedyITCase.testGreedyZeroOrMoreBeforeOptional2` is due to that proceed. The whole reason behind the additional ignoreState was for it not to have the `PROCEED` transition as it creates additional computationStates with the `sinkState`. I think for the greedy to work we need to put customized ignoreCondition into the `sinkState` of a greedy Pattern. What do you think?